odl/tomo/backends/astra_cuda.py

```diff
@@ -0,0 +1,259 @@
+
+try:
+    import astra
+    ASTRA_CUDA_AVAILABLE = astra.astra.use_cuda()
+except ImportError:
+    ASTRA_CUDA_AVAILABLE = False
+
+# Internal
+from odl.discr.lp_discr import DiscreteLp, DiscreteLpVector
+from odl.tomo.backends.astra_setup import (
+    astra_projection_geometry, astra_volume_geometry, astra_projector,
+    astra_data, astra_algorithm)
+from odl.tomo.geometry import Geometry
+from odl.tomo.geometry import (
+    Parallel2dGeometry, FanFlatGeometry, Parallel3dGeometry,
+    HelicalConeFlatGeometry)
+
+__all__ = ('astra_cuda_forward_projector', 'astra_cuda_back_projector',
+           'ASTRA_CUDA_AVAILABLE')
+
+
+# TODO: use context manager when creating data structures
+# TODO: is magnification scaling at the right place?
+
+def astra_cuda_forward_projector(vol_data, geometry, proj_space, out=None):
+    """Run an ASTRA forward projection on the given data using the GPU.
+
+    Parameters
+    ----------
+    vol_data : `DiscreteLpVector`
+        Volume data to which the projector is applied
+    geometry : `Geometry`
+        Geometry defining the tomographic setup
+    proj_space : `DiscreteLp`
+        Space to which the calling operator maps
+    out : `DiscreteLpVector`, optional
+        Vector in the projection space to which the result is written. If
+        `None` creates an element in the projection space ``proj_space``
+
+    Returns
+    -------
+    out : ``proj_space`` element
+        Projection data resulting from the application of the projector
+    """
+    if not isinstance(vol_data, DiscreteLpVector):
+        raise TypeError('volume data {!r} is not a `DiscreteLp.Vector` '
+                        'instance.'.format(vol_data))
+    if not isinstance(geometry, Geometry):
+        raise TypeError('geometry  {!r} is not a `Geometry` instance.'
+                        ''.format(geometry))
+    if vol_data.ndim != geometry.ndim:
+        raise ValueError('dimensions {} of volume data and {} of geometry '
+                         'do not match.'
+                         ''.format(vol_data.ndim, geometry.ndim))
+    if not isinstance(proj_space, DiscreteLp):
+        raise TypeError('projection space {!r} is not a `DiscreteLp` '
+                        'instance.'.format(proj_space))
+    if out is not None:
+        if not isinstance(out, DiscreteLpVector):
+            raise TypeError('out {} is neither `None` nor a '
+                            '`DiscreteLpVector` instance'.format(out))
+
+    ndim = vol_data.ndim
+
+    # Create astra geometries
+    vol_geom = astra_volume_geometry(vol_data.space)
+    proj_geom = astra_projection_geometry(geometry)
+
+    # Create ASTRA data structures
+
+    # In the case dim == 3, we need to swap axes, so can't perform the FP
+    # in-place
+    if out is None and ndim == 2:
+        out = proj_space.element()
+
+    vol_id = astra_data(vol_geom, datatype='volume', data=vol_data)
+    sino_id = astra_data(proj_geom, datatype='projection', data=out,
```

Comment:
Here data does not make sense in the case of 3 dimensions, in that case we do a copy anyway (further down).